From c6758fa018078daf59f9781a1eab4bc4eb8cac05 Mon Sep 17 00:00:00 2001
From: Zhang Qiao <zhangqiao22@huawei.com>
Date: Fri, 17 Jun 2022 01:52:26 +0000
Subject: [PATCH 2595/4585] [Huawei] Revert "[Backport] sched: Fix yet more
 sched_fork() races"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186973

--------------------------------

This reverts commit 91dc59c2c7ad3fd1ac5378d5e80f9d9180edc071.
the patch("sched: Fix yet more sched_fork()") may be cause a
process sleep at cgroup_post_fork()->freezer_fork() while
taking group_threadgroup_rwsem lock long time, it cause a
problem that other tasks will wait while fork child process
and the system will stall.

Signed-off-by: Zhang Qiao <zhangqiao22@huawei.com>
---
 include/linux/sched/task.h |  4 ++--
 kernel/fork.c              | 13 +------------
 kernel/sched/core.c        | 23 +++++++----------------
 3 files changed, 10 insertions(+), 30 deletions(-)

diff --git a/include/linux/sched/task.h b/include/linux/sched/task.h
index cea4bdfd0f05..fa75f325dad5 100644
--- a/include/linux/sched/task.h
+++ b/include/linux/sched/task.h
@@ -55,8 +55,8 @@ extern asmlinkage void schedule_tail(struct task_struct *prev);
 extern void init_idle(struct task_struct *idle, int cpu);
 
 extern int sched_fork(unsigned long clone_flags, struct task_struct *p);
-extern void sched_cgroup_fork(struct task_struct *p, struct kernel_clone_args *kargs);
-extern void sched_post_fork(struct task_struct *p);
+extern void sched_post_fork(struct task_struct *p,
+			    struct kernel_clone_args *kargs);
 extern void sched_dead(struct task_struct *p);
 
 void __noreturn do_task_dead(void);
diff --git a/kernel/fork.c b/kernel/fork.c
index f932fd8b4bd5..3b67fb1fd3c4 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -2248,17 +2248,6 @@ static __latent_entropy struct task_struct *copy_process(
 	if (retval)
 		goto bad_fork_put_pidfd;
 
-	/*
-	 * Now that the cgroups are pinned, re-clone the parent cgroup and put
-	 * the new task on the correct runqueue. All this *before* the task
-	 * becomes visible.
-	 *
-	 * This isn't part of ->can_fork() because while the re-cloning is
-	 * cgroup specific, it unconditionally needs to place the task on a
-	 * runqueue.
-	 */
-	sched_cgroup_fork(p, args);
-
 	/*
 	 * From this point on we must avoid any synchronous user-space
 	 * communication until we take the tasklist-lock. In particular, we do
@@ -2366,7 +2355,7 @@ static __latent_entropy struct task_struct *copy_process(
 		fd_install(pidfd, pidfile);
 
 	proc_fork_connector(p);
-	sched_post_fork(p);
+	sched_post_fork(p, args);
 	cgroup_post_fork(p, args);
 	perf_event_fork(p);
 
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 1b8bcedc8643..5e6d7ce8959c 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -3322,7 +3322,6 @@ int sched_fork(unsigned long clone_flags, struct task_struct *p)
 
 	init_entity_runnable_average(&p->se);
 
-
 #ifdef CONFIG_SCHED_INFO
 	if (likely(sched_info_on()))
 		memset(&p->sched_info, 0, sizeof(p->sched_info));
@@ -3338,23 +3337,18 @@ int sched_fork(unsigned long clone_flags, struct task_struct *p)
 	return 0;
 }
 
-void sched_cgroup_fork(struct task_struct *p, struct kernel_clone_args *kargs)
+void sched_post_fork(struct task_struct *p, struct kernel_clone_args *kargs)
 {
 	unsigned long flags;
+#ifdef CONFIG_CGROUP_SCHED
+	struct task_group *tg;
+#endif
 
-	/*
-	 * Because we're not yet on the pid-hash, p->pi_lock isn't strictly
-	 * required yet, but lockdep gets upset if rules are violated.
-	 */
 	raw_spin_lock_irqsave(&p->pi_lock, flags);
 #ifdef CONFIG_CGROUP_SCHED
-	if (1) {
-		struct task_group *tg;
-		tg = container_of(kargs->cset->subsys[cpu_cgrp_id],
-				  struct task_group, css);
-		tg = autogroup_task_group(p, tg);
-		p->sched_task_group = tg;
-	}
+	tg = container_of(kargs->cset->subsys[cpu_cgrp_id],
+			  struct task_group, css);
+	p->sched_task_group = autogroup_task_group(p, tg);
 #endif
 	rseq_migrate(p);
 	/*
@@ -3365,10 +3359,7 @@ void sched_cgroup_fork(struct task_struct *p, struct kernel_clone_args *kargs)
 	if (p->sched_class->task_fork)
 		p->sched_class->task_fork(p);
 	raw_spin_unlock_irqrestore(&p->pi_lock, flags);
-}
 
-void sched_post_fork(struct task_struct *p)
-{
 	uclamp_post_fork(p);
 }
 
-- 
2.17.1

